src/backends/unix/system.c

```diff
@@ -548,27 +548,92 @@ Sys_IsFile(const char *path)
 	return false;
 }
 
-char *
-Sys_GetHomeDir(void)
+#ifdef USE_XDG
+static char *
+GetXDGPath(const char *xdg)
 {
-	static char gdir[MAX_OSPATH];
-	char *home;
+	char* buffer = calloc(MAX_OSPATH, sizeof(char));
+	if (!buffer) {
+		goto fail;
+	}
 
-	home = getenv("HOME");
+	const char* env = getenv(xdg);
+	const char* fmt = "%s/%s/";
 
-	if (!home)
-	{
-		return NULL;
+	if (!env) {
+		env = getenv("HOME");
+		if (!env) {
+			goto fail;
+		}
+
+		if (strcmp(xdg, "XDG_CONFIG_HOME")==0) {
+			fmt = "%s/.config/%s/";
+		} else if (strcmp(xdg, "XDG_DATA_HOME")==0) {
+			fmt = "%s/.local/share/%s/";
+		} else if (strcmp(xdg, "XDG_STATE_HOME")==0) {
+			fmt = "%s/.local/state/%s/";
+		} else {
+			Sys_Error("%s: unexpected directory %s", __func__, xdg);
+		}
 	}
 
+	Com_sprintf(buffer, MAX_OSPATH, fmt, env, cfgdir);
+	return buffer;
+
+fail:
+	free(buffer);
+	return NULL;
+}
+#endif
+
+char *
+Sys_GetHomeDir()
+{
+	static char dir[MAX_OSPATH];
+
+	if (!dir[0]) {
+		const char* home = getenv("HOME");
+
+		if (!home) {
+			// uh-oh
+			return NULL;
+		}
+
 #ifndef __HAIKU__
-	Com_sprintf(gdir, sizeof(gdir), "%s/%s/", home, cfgdir);
-#else
-	Com_sprintf(gdir, sizeof(gdir), "%s/config/settings/%s", home, cfgdir);
+		if (strcmp(cfgdir, CFGDIRNAME)!=0) {
+			// custom cfgdir was set by the user: ~/{cfgdir}
+			Com_sprintf(dir, MAX_OSPATH, "%s/%s/", home, cfgdir);
+			goto dirset;
+		}
+
+		// hidden dir: ~/.{CFGDIRNAME_SHORT}
+		Com_sprintf(dir, MAX_OSPATH, "%s/.%s/", home, CFGDIRNAME_SHORT);
+
+#ifdef USE_XDG
+		if (Sys_IsDir(dir)) {
+			goto dirset;
```

Comment:
Could just write message about backward compatibility and return dir, if I correctly understand logic that code see existing old config directory and reuse it?